sled-agent/src/swap_device.rs

```diff
@@ -0,0 +1,493 @@
+///   1. On startup, check if there is a swap device. If one exists, we are done.
+///      Swap devices do not persist across reboot by default, so if a device
+///      already exists, this isn't our first time starting after boot. The
+///      device may be in use. Changes to how the swap device is setup, should we
+///      decide to do that, will be across reboots (as this is how sled-agent is
+///      upgraded), so we will get a shot to make changes across upgrade.
+///   2. If there is no swap device, check for a zvol at the known path on the
+///      M.2 that we booted from. If we find such a zvol, delete it.
+///   3. Create an encrypted zvol with a randomly generated key that is
+///      immediately discarded.
+///   4. Add the zvol as a swap device with swapctl(2).
+///
+/// Note that this introduces a sled-agent upgrade consideration if we ever
+/// choose to change how we set up the device. A configured swap device does not
+/// persist across reboot by default, but a zvol does. Thus, if the well known
+/// path for the zvol ever changes, we will need to at least support a window
+/// where we check for both the previously well-known path and the new
+/// configuration.
+pub(crate) fn ensure_swap_device(
+    log: &slog::Logger,
+    boot_zpool_name: &illumos_utils::zpool::ZpoolName,
+    size_gb: u32,
+) -> Result<(), SwapDeviceError> {
+    assert!(size_gb > 0);
+
+    let devs = swapctl::list_swap_devices()?;
+    if devs.len() > 0 {
+        if devs.len() > 1 {
+            // This should really never happen unless we've made a mistake, but it's
+            // probably fine to have more than one swap device. Thus, don't panic
+            // over it, but do log a warning so there is evidence that we found
+            // extra devices.
+            //
+            // Eventually, we should probably send an ereport here.
+            warn!(
+                log,
+                "Found multiple existing swap devices on startup: {:?}", devs
+            );
+        } else {
+            info!(log, "Swap device already exists: {:?}", devs);
+        }
+
+        return Ok(());
+    }
+
+    let swap_zvol = format!("{}/{}", boot_zpool_name, "swap");
+    if zvol_exists(&swap_zvol)? {
+        info!(log, "swap zvol \"{}\" aleady exists; destroying", swap_zvol);
+        zvol_destroy(&swap_zvol)?;
+        info!(log, "swap zvol \"{}\" destroyed", swap_zvol);
+    }
+
+    create_encrypted_swap_zvol(log, &swap_zvol, size_gb)?;
+
+    // The process of paging out using block I/O, so use the "dsk" version of
+    // the zvol path (as opposed to "rdsk", which is for character/raw access.)
+    let blk_zvol_path = format!("/dev/zvol/dsk/{}", swap_zvol);
+    info!(log, "adding swap device: swapname=\"{}\"", blk_zvol_path);
+
+    // Specifying 0 length tells the kernel to use the size of the device.
+    swapctl::add_swap_device(blk_zvol_path, 0, 0)?;
+
+    Ok(())
+}
+
+// Check whether the given zvol exists.
```

Comment:
I'd make this comment and the other private-function comments into doc comments even though the functions aren't `pub` (my dev environment, and I assume other similar IDEs, will only show these comments on hover if they're doc comments).

Reply:
done in b739a72